main.cpp

```diff
@@ -0,0 +1,121 @@
+#include <iostream>
+#include <stdio.h>
+#include <stdlib.h>
+extern "C" {
+#include "mmio.h"
+}
+
+void matrix_vector(int rows, int cols, double* A, double* v, double* out) {
+    for(int r=0; r<rows; r++) {
+        double rowSum = 0;
+        for(int c=0; c<cols; c++) {
+            rowSum += *(A+cols*r+c) * *(v+c);
+        }
+        *(out+r) = rowSum;
+    }
+}
+
+void matrix_matrix(int Arows, int Acols, int Bcols, double* A, double* B, double* out) {
+    for(int c=0; c<Bcols; c++) {
+        // extract column from B
+        double v[Acols];
+        for(int i=0; i<Acols; i++) {
+            int iB = Bcols*i+c;
+            v[i] = B[iB];
+        }
+        // deposit into columns
+        double outCol[Arows];
+        matrix_vector(Arows, Acols, A, v, outCol);
+        for(int i=0; i<Arows; i++) {
+            int iOut = Bcols*i+c;
+            out[iOut] = outCol[i];
+        } 
+    }
+}
+
```

Comment:
Since the functions are small, they may go inside the main cpp file although generally, it is not ideal.